
package robominions;

option java_package = "au.id.katharos.robominions.api";
option java_outer_classname = "RobotApi";


import "materials.proto";


message RobotRequest {
  optional string name = 1;
  optional int32 key = 2;

  // Pick one, cannot be both an action and a read.
  optional RobotReadRequest read_request = 3;
  optional RobotActionRequest action_request = 4;
}

message Coordinate {
  required int32 x = 1;
  required int32 y = 2;
  required int32 z = 3;
}

message WorldLocation {
  enum Direction {
    UP = 0;
    DOWN = 1;
    LEFT = 2;
    RIGHT = 3;
    FORWARD = 4;
    BACKWARD = 5;
    EAST = 6;
    WEST = 7;
    NORTH = 8;
    SOUTH = 9;
  }
 
  optional Coordinate absolute_location = 1;
  optional Direction direction = 2;
  // TODO: Consider adding relative location
  // e.g. (0, 3, 2) for 3 blocks up, 2 across 
}

message RobotReadRequest {
  enum Entity {
  	SELF = 0;
    OWNER = 1;
  }

  // These return zero or more materials
  optional WorldLocation identify_material = 1;
  
  // These return zero or more locations
  optional Material locate_material_nearby = 2;
  optional Entity locate_entity = 3;
  
  // TODO: Query the solidness of a block
}

message RobotActionRequest {
  optional WorldLocation.Direction move_direction = 2;
  optional WorldLocation.Direction turn_direction = 3;
  optional WorldLocation.Direction mine_direction = 4;
  
  optional WorldLocation.Direction place_direction = 5;
  optional Material place_material = 6;
}

message LocationResponse {
  repeated WorldLocation locations = 1;
}

message MaterialResponse {
  repeated Material materials = 1;
}

message RobotResponse {
  optional int32 key = 3;
  optional bool success = 1;
  optional ErrorMessage error_message = 2;
  
  optional LocationResponse location_response = 4;
  optional MaterialResponse material_response = 5;
}

message ErrorMessage {
  enum Reason {
  	UNKNOWN = 0;
  	SERVER_ERROR = 1;
  	UNREADABLE_REQUEST = 2;
  	INVALID_REQUEST = 3;
  	BLOCK_COLLISION = 4;
  	OUTSIDE_OF_WORLD = 5;
  	ROBOT_DOES_NOT_EXIST = 6;
  	BLOCK_IS_NOT_REACHABLE = 7;
  	BLOCK_IS_NOT_VISIBLE = 8;
  	NOT_IMPLEMENTED = 9;
  }
  enum Action {
  	FAIL_ACTION = 0;
  	RETRY_ACTION = 1;
  	EXIT_CLIENT = 2;
  }
  optional Reason reason = 1;
  optional Action action = 2;
  optional string message = 3;
}
